bundles/binding/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/MyqData.java

```diff
@@ -0,0 +1,208 @@
+				webSite, appId, getSecurityToken());
+
+		JsonNode data = request("GET", url, null, null, true);
+
+		return new GarageDoorData(data);
+	}
+
+	/**
+	 * Validates Username and Password then saved sercurityToken to a variable
+	 */
+	private void login() throws InvalidLoginException, IOException {
+		logger.debug("attempting to login");
+		String url = String
+				.format("%s/api/user/validate?appId=%s&SecurityToken=null&username=%s&password=%s",
+						webSite, appId, userName, password);
+
+		JsonNode data = request("GET", url, null, null, true);
+		LoginData login = new LoginData(data);
+		sercurityToken = login.getSecurityToken();
+	}
+
+	/**
+	 * Send Command to open/close garage door opener with MyQ API Returns false
+	 * if return code from API is not correct or connection fails
+	 * 
+	 * @param deviceID
+	 *            MyQ deviceID of Garage Door Opener.
+	 * @param state
+	 *            Desired state to put the door in, 1 = open, 0 = closed
+	 */
+	public void executeGarageDoorCommand(int deviceID, int state)
+			throws InvalidLoginException, IOException {
+		String message = String.format("{\"ApplicationId\":\"%s\","
+				+ "\"SecurityToken\":\"%s\"," + "\"MyQDeviceId\":\"%d\","
+				+ "\"AttributeName\":\"desireddoorstate\","
+				+ "\"AttributeValue\":\"%d\"}", appId, sercurityToken,
+				deviceID, state);
+		String url = String
+				.format("%s/api/v4/deviceattribute/putdeviceattribute?appId=%s&SecurityToken=%s",
+						webSite, appId, getSecurityToken());
+
+		request("PUT", url, message, "application/json", true);
+	}
+
+	/**
+	 * Returns the currently cached security token, this will make a call to
+	 * login if the token does not exist.
+	 * 
+	 * @return The cached security token
+	 * @throws IOException
+	 * @throws InvalidLoginException
+	 */
+	private String getSecurityToken() throws IOException, InvalidLoginException {
+		if (sercurityToken == null) {
+			login();
+		}
+		return sercurityToken;
+	}
+
+	/**
+	 * Make a request to the server, optionally retry the call if there is a
+	 * login issue. Will throw a InvalidLoginExcpetion if the account is
+	 * invalid, locked or soon to be locked.
+	 * 
+	 * @param method
+	 *            The Http Method Type (GET,PUT)
+	 * @param url
+	 *            The request URL
+	 * @param payload
+	 *            Payload string for put operations
+	 * @param payloadType
+	 *            Payload content type for put operations
+	 * @param retry
+	 *            Retry the attempt if our session key is not valid
+	 * @return The JsonNode representing the response data
+	 * @throws IOException
+	 * @throws InvalidLoginException
+	 */
+	private synchronized JsonNode request(String method, String url,
+			String payload, String payloadType, boolean retry)
+			throws IOException, InvalidLoginException {
+
+		logger.debug("Requsting URL {}", url);
+
+		String dataString = executeUrl(method, url, header,
+				payload == null ? null : IOUtils.toInputStream(payload),
+				payloadType, 5000);
```

Comment:
Is there any chance at all that a 5-second timeout will in some case be insufficient?  If so, could this be a binding config value in openhab.cfg / myq.cfg ?